gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Client.java

```diff
@@ -561,16 +568,56 @@ protected void initializeImplementation() {
 
         private Consumer<Host> initializeConnectionSetupForHost = host -> {
             try {
-                // hosts that don't initialize connection pools will come up as a dead host
+                // hosts that don't initialize connection pools will come up as a dead host.
                 hostConnectionPools.put(host, new ConnectionPool(host, ClusteredClient.this));
 
-                // added a new host to the cluster so let the load-balancer know
+                // hosts are not marked as available at cluster initialization, and are made available here instead.
+                host.makeAvailable();
+
+                // added a new host to the cluster so let the load-balancer know.
                 ClusteredClient.this.cluster.loadBalancingStrategy().onNew(host);
             } catch (RuntimeException ex) {
-                final String errMsg = "Could not initialize client for " + host;
-                throw new RuntimeException(errMsg, ex);
+                throw new RuntimeException(String.format("Could not initialize client for %s.", host), ex);
             }
         };
+
+        private void handleUnavailableHosts(List<Host> unavailableHosts, ExecutorService hostExecutor) {
+            // start the re-initialization attempt for each of the unavailable hosts through Host.makeUnavailable().
+            try {
+                CompletableFuture.allOf(unavailableHosts.stream()
+                                .map(host -> CompletableFuture.runAsync(() -> host.makeUnavailable(this::tryReInitializeHost), hostExecutor))
+                                .toArray(CompletableFuture[]::new))
+                        .join();
+            } catch (CompletionException ex) {
+                logger.error("", (ex.getCause() == null) ? ex : ex.getCause());
+            } finally {
+                hostExecutor.shutdown();
+            }
+        }
+
+        /**
+         * Attempt to re-initialize the {@link Host} that was previously marked as unavailable.  This method gets called
+         * as part of a schedule in {@link Host} to periodically try to re-initialize.
+         */
+        public boolean tryReInitializeHost(final Host host) {
+            logger.debug("Trying to re-initiate host connection pool on {}", host);
+
+            try {
+                // hosts that don't initialize connection pools will come up as a dead host
+                hostConnectionPools.put(host, new ConnectionPool(host, ClusteredClient.this));
+
+                // added a new host to the cluster so let the load-balancer know
+                ClusteredClient.this.cluster.loadBalancingStrategy().onNew(host);
+
+                // mark this host as available
+                host.makeAvailable();
```

Comment:
this is already done in `reconnected()` function called in `makeUnavailable()`

Reply:
re-using `initializeConnectionSetupForHost()` as noted below.